docs/architecture/adr-004-coroutine-effect-system.md

```diff
@@ -0,0 +1,303 @@
+# ADR 004: Coroutine-Based Effect System for Consensus
+
+## Changelog
+
+* 18-03-2025: Initial version
+
+## Status
+
+Accepted
+
+## Context
+
+The Malachite consensus engine needs to interact with its environment (network, storage, cryptography, application logic) at specific points during execution.
```

Comment:
So, the engine here is the state-machine, driver, what?

I thought that the engine included some network implementation etc.

Reply:
I agree. Might be best to say explicitly here "core library for Tendermint".